### ISSUE TITLE
Question about signatures

### ISSUE THREAD

**Issue by eurobank** [CIDRAM/CIDRAM#477]
@Maikuolan How reliable are the signatures, IPv4, Spammy ISPs and proxies VPN? 

Who is collecting them and from where? I ask because i get large scale attacks by means of time (one day maybe) and none of the IPs are in any signature. The attacks are logged and blocked by Wordpress security plugins.

**Comment by Maikuolan** [comment]
> Who is collecting them and from where?

Mostly collected myself, via my own honeypots and through my own research, though sometimes via suggestions and data supplied from other CIDRAM users via email, occasionally (but rarely) suggestions given to me via discussions and messaging at various forums and online communities I visit.

> How reliable are the signatures

Depends on perspective.

On one hand, I don't add data arbitrarily, so for any data added to the signature files, there's always a story behind it, evidence I've seen of attacks, spam, or some other kind of unwanted traffic from the networks or addresses which that data concerns. So, the likelihood of something being blocked for seemingly no reason at all is pretty much zero.

On the other hand, I'm only one person, my honeypots don't cover the entirety of the internet, and the number of attacks, malware, spammers, and everything else on the internet, already a ridiculously huge problem, is only getting worse by the day, with no signs of getting better any time soon. In other words, no matter how closely I monitor my honeypots, no matter how much research I do, even if I was to spend 24 hours a day doing it, the likelihood of me being able to detect or anticipate everything from everywhere, and therefore being able to add new signatures to the signature files to deal with those detected, anticipated attacks, spam, etc, is also pretty much zero.

An additional problem, compounding that problem, is that, the larger the signature files get over time, the more time that's required to cycle over all old data, to keep it up-to-date, to remove any outdated, no longer valid and no longer needed signatures (and that's especially true for the spammy ISPs signature files, as we don't generally want to list data from ISPs for any longer than is absolutely necessary, as IP allocations for domestic usage tend to change much more frequently than IP allocations for hosting services, cloud providers, etc, and the risk of a dirty IP address being abandoned by an attacker and subsequently reallocated to someone innocent is something to always consider when cycling through the ISPs signature files). Because I'm the only person updating them, as they grow over time, the longer it takes to cycle through old data, the greater the risk that stale, no longer wanted data remains in the signature files, which means, the greater the risk of that data causing false positives to occur.

So, how to deal with that? I just do what I can, and hope that the modules (e.g., the AbuseIPDB module, the Stop Forum Spam module, the Project Honeypot module, the bot user agents module, etc) can fill most of the gaps left behind by the signature files, and for anything else that still manages to get through, leave it up to configuring the affected installations appropriately for the needs of the websites they protect, and up to auxiliary rules.

It might've been different a few years ago, but as it stands right now (the current size of the signature files, when the current least recently updated data was last updated, etc):
* I wouldn't risk using the spammy ISPs signature files unless CAPTCHAs are also enabled, with reasonably permissive configuration behind it, to reduce the risk that, in the event of a false positive, that the person accidentally blocked doesn't remain locked out (in fact, for the various websites that I help maintain, where I have CIDRAM installed and running, I only use the spammy ISPs signature files at the low-traffic websites, and at websites where such false positives wouldn't have much consequence, e.g., potential loss of customers to online stores, things like that). With CAPTCHAs, or some other way to get around the block, it should be fine, but I wouldn't risk it without it.
* The risk of false positives should be much less with the other signature files, and it should generally be beneficial to make sure they're active. In particular, unless I notice some kind of unusual problem with them at the installation in question, I wouldn't want to be without the standard IPv4/IPv6 cloud signature files. There's a *lot* of unwanted crap that'll be blocked as a result of having those signature files enabled, and.. that point hasn't really ever changed since CIDRAM started. However, that said..
* ..Due to the overall situation of the internet seemingly getting worse over time (in regard to unwanted traffic, spam, attacks, hacking, etc), and due the size of the team responsible for updating the signature files not getting any bigger (i.e., it's still just me doing it alone), using the signature files in isolation, without the backing of any modules, auxiliary rules, etc, is likely going to inhibit CIDRAM's effectiveness quite a bit, likely going to see a lot of that unwanted traffic still getting through. I'm not sure I'd want to have an installation not use any of the modules anymore, and in fact, this past year and the past few years, most of the more severe attacks, the most dangerous traffic, hack attempts, etc blocked by CIDRAM, have been due to modules, without any help from the signature files whatsoever. So..
* ..Modules could arguably be more useful than signature files in that sense. However, I wouldn't use modules in isolation, either. One of the biggest reasons for that? In the execution chain for CIDRAM, signature files are processed prior to modules, and for any requests already blocked by signature files, we don't need the modules to bother performing any API lookups, which means, for any requests already blocked by signature files, we're not wasting lookup quota. And, although modules seem to be blocking more of the more dangerous traffic than signature files per my own more recent experience, the sheer raw number of requests being blocked by signature files still seems to be much higher than the number of requests being blocked by modules alone, which tells me that the presence of the signature files is helping to minimise how much of my lookup quota is being used per day at my CIDRAM installations. Which is a good thing, because the moment my lookup quota is exceeded, those modules won't be able to perform any new lookups until my quota resets at the end of the day, meaning they won't be able to block anything they encounter which hasn't already been blocked that day or which isn't being tracked by IP tracking at the time, isn't currently blocked by signature files, etc. In the event that signature files aren't present, I could even foresee a potential attacker sending a few thousand arbitrary, innocuous requests from all different addresses, maybe via some kind of VPN or hosting provider which allows requests to be sent from a rotating pool of different IPs, all for the sake of eating up the daily lookup quota, and then, after the daily lookup quota has been exceeded, starting the real attack with the real intended payload from some other, new IPs. (Well.. though I say that.. most attacks nowadays are totally automated, coming from bots which attack several thousand websites at once, so the likelihood of encountering something so narrowly targeted, so fine-tuned to any specific website is also extremely low, but.. theoretically possible, so still greater than zero, at least).
* Above points have covered the cloud signatures and the ISP signatures. As for VPN and proxy signatures? Eh.. so-so. If you're running a website which isn't particularly amicable towards the idea of being accessed via VPNs or proxies, I'd say the VPN and proxy signature files are still worth using. If, however, you're okay with the idea of access via VPNs or proxies, it might be better to not use those signature files. They do block some unwanted traffic, but the number of signatures they provide is very small compared to the other signature files, and the benefit provided is similarly also very small. The risk of false positives in the case of those signature files, I would suggest, also strongly correlates with how much you actually do/don't want to see VPNs or proxies at your website. That in mind, my recommendation about whether use them or not would need to be case-by-case, I think.

Bogon signature files.. Likelihood of blocking anything is always going to be very small, for the same reason that keeping those signature files active is going to be a good idea in most cases: We don't typically expect to see traffic of any kind from bogon addresses, so seeing such traffic, typically, doesn't indicate anything good. Likelihood of bogon signatures blocking any actual users is pretty much zero. However, for any websites which utilise cronjobs, daemons, or something else at the same machine as their HTTP server, which sends internal requests between the HTTP server and whatever else is at that same machine, the bogon signatures could potentially cause some false positives, in that such internal requests could, potentially, fall within those typically unexpected IP ranges. (But such problems, typically, are very easy to diagnose and fix).

The "non-blocking" signatures, as per their name, don't actually block anything at all, but are intended to populate the "profile" for certain kinds of requests, which serves to feed some additional data to auxiliary rules which can be used for fine-tuning certain kinds of custom auxiliary rules. Totally useless for any user which doesn't utilise auxiliary rules at all, but potentially useful in some cases for any user which does. But, probably not needed in most cases.

**Comment by eurobank** [comment]
I didn't imagine that you were collecting most of the signatures. A big project by any means.

So, now, even more than before, i believe that you should think two possible future improvements.

1. Find some way for us to GIVE BACK our results, in a verified way, I dunno how many we are, still some verified data could be shared.
2. Seriously think to use something like crowdsec or some alternative
